src/Composer/Satis/Builder/PackagesBuilder.php

```diff
@@ -126,29 +134,71 @@ private function pruneIncludeDirectories()
     private function dumpPackageIncludeJson(array $packages, $includesUrl, $hashAlgorithm = 'sha1')
     {
         $filename = str_replace('%hash%', 'prep', $includesUrl);
-        $path = $this->outputDir . '/' . ltrim($filename, '/');
+        $path = $tmpPath = $this->outputDir . '/' . ltrim($filename, '/');
 
         $repoJson = new JsonFile($path);
-        $repoJson->write(array('packages' => $packages));
+        $contents = $repoJson->encode(array('packages' => $packages)) . "\n";
 
-        $hash = hash_file($hashAlgorithm, $path);
+        $hash = hash($hashAlgorithm, $contents);
 
         if (strpos($includesUrl, '%hash%') !== false) {
             $this->writtenIncludeJsons[] = array($hash, $includesUrl);
             $filename = str_replace('%hash%', $hash, $includesUrl);
-            rename(
-                $path,
-                $path = $this->outputDir . '/' . ltrim($filename, '/')
-            );
+            if (file_exists($path = $this->outputDir . '/' . ltrim($filename, '/'))) {
+                // When the file exists, we don't need to override it as we assume,
+                // the contents satisfy the hash
+                $path = null;
+            }
+        }
+        if ($path) {
+            $this->writeToFile($path, $contents);
+            $this->output->writeln("<info>wrote packages to $path</info>");
         }
-        $this->output->writeln("<info>wrote packages to $path</info>");
 
         return array(
             $filename => array($hashAlgorithm => $hash)
         );
     }
 
     /**
+     * Write to a file
+     *
+     * @param string $path
+     * @param string $contents
+     */
+    private function writeToFile($path, $contents)
```

Comment:
i think this method needs explanation why it is here. what for the retries are there. it also looks like copy paste from some other library. imho should be noted so (via `@see URL`  perhaps)